src/utils/messaging-outbox/src/consumers/message_consumers_mediator_utils.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+use dill::{Builder, BuilderExt, Catalog, TypecastBuilder};
+use futures::{StreamExt, TryStreamExt};
+use internal_error::{InternalError, ResultIntoInternal};
+
+use super::{ConsumerFilter, MessageConsumer, MessageConsumerMeta, MessageConsumptionDurability};
+use crate::{Message, MessageConsumerT, MessageConsumersMediator, MessageSubscription};
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[tracing::instrument(level = "debug", skip_all, fields(%content_json))]
+pub async fn consume_deserialized_message<'a, TMessage: Message + 'static>(
+    catalog: &Catalog,
+    consumer_filter: ConsumerFilter<'a>,
+    content_json: serde_json::Value,
+) -> Result<(), InternalError> {
+    let message = serde_json::from_value::<TMessage>(content_json).int_err()?;
+
+    let consumers = match consumer_filter {
+        ConsumerFilter::AllConsumers => all_consumers_for::<TMessage>(catalog),
+        ConsumerFilter::BestEffortConsumers => best_effort_consumers_for::<TMessage>(catalog),
+        ConsumerFilter::SelectedConsumer(consumer_name) => {
+            particular_consumers_for::<TMessage>(catalog, consumer_name)
+        }
+    };
+
+    let consumption_tasks = consumers
+        .into_iter()
+        .map(|consumer| (consumer, message.clone()));
```

Comment:
Can we pass message by ref to avoid cloning it per consumer?

Reply:
yes